data/org.freedesktop.impl.portal.ScreenCast.xml

```diff
@@ -123,6 +154,44 @@
               'multiple' was set to 'true' as part of the SelectSources method.
             </para></listitem>
           </varlistentry>
+          <varlistentry>
+            <term>persist_mode u</term>
+            <listitem><para>
+              Portal implementations can optionally offer ways to "reduce" the
+              persist mode, for example by explicitly asking the user. See the
+              'persist_mode' option of SelectSources for a list of possible
+              values.
+
+              If no persist mode is returned by the Start request, it is assumed
+              to be the same persist mode received during SelectSources.
+
+              This option was added in version 4 of this interface.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>restore_data v</term>
+            <listitem><para>
+              The data to be restored. This is the data that a portal
+              implementation sent in reaction to transient or persistent
+              modes.
+
+              This data will be stored in the permission store if the effective
+              persist mode is 2, and may be passed in a future as part of the
+              SelectSources() method call. The portal backend should store enough
+              information in 'restore_data' to be able to restore the session
+              later. The portal backend can deal with unavailable sources however
+              they see fit.
+
+              Data generated by a portal implementation must be consumable by
+              the same portal implementation, but not necessarily by others.
+              Thus, it is important to properly handle 'restore_data' not being
+              in a valid format. This may happen when, e.g., the user switched
+              to another desktop environment, or is using a different portal
```

Comment:
Does this mean, that changing portal implementation would drop a stored permission an implementation doesn't understand, or would this result in multiple entries, which are send sequentially until the implementation does find a valid one?
Would it be possible to register the namespace of the portal implementation with it?

Reply:
> Does this mean, that changing portal implementation would drop a stored permission an implementation doesn't understand

This.

>  Would it be possible to register the namespace of the portal implementation with it?

I don't think forcing any particular format here would be helpful. Giving portal frontends the flexibility to store whatever data they want is the key part of this; it allows frontends to implement whatever interaction they see fit. Each portal frontend should check if the variant they're receiving is in a format they recognize, then parse this variant and check if the values are expected.

[For the GNOME implementation](https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome/-/merge_requests/14), we went with `(suxxa(uus))` as the variant format, because we need to store: `("GNOME", version, creation time, last used time, array of (stream id, stream type, stream recovery string))`. The "creation time" and "last used time" fields will be used by GNOME Settings to show information about this permission; "stream recovery string" is Mutter-specific.

Reply:
> I don't think forcing any particular format here would be helpful. Giving portal frontends the flexibility to store whatever data they want is the key part of this;

I don't want to force a full spec on what can be stored, but I usually like structures with a fixed format and one entry for arbitrary data. I have no experience with dbus and as such i can't really offer a working solution, but would it possible to save the path of the portal `org.freedesktop.impl.portal.{gnome,wlr,...}` in some specified entry of the permission store and the permission store will only send entries matching to the token and the impl? Another idea would be to specify that the first entry in the permission store has to be the portal name, such that all data would start with "GNOME", "WLROOTS", ... and this would ease parsing if a valid format is present.

DE independent permissions aren't possible at all with the current state, but it would be nice for people using multiple DE's to have the permanent settings survive a switch. 

This probably would take a "keep token" response to the client and managing multiple tokens and would make the whole thing much more complicated... 

So tldr: A standardized first entry in the variant format would be nice to avoid collitions, everything else is just nice to have if possible in a reasonable way.

Reply:
> DE independent permissions aren't possible at all with the current state

This is currently possible, it's just discouraged. I'd prefer that each portal implementation focuses on itself and how to provide the best experience they can. But any portal implementation can check the variant type and see if it matches any other implementation's format and load it. You don't need an extra string for that.

Reply:
> This is currently possible, it's just discouraged.

How with non-standardized restore_data? 

Another idea: Should the data be only used by the portal, or should third applications (like flatseal) be allowed to edit those entries? In this case the application has to extinguish between datas from different portals.

Reply:
> How with non-standardized restore_data?

```c
if (g_variant_check_format_string (data, "(suxxa(uus))", FALSE))
  {
      /* restore GNOME session */
  }
else if (g_variant_check_format_string (data, "(foo{bar})", FALSE))
  {
      /* restore someone else's session */
  }
```

Reply:
Ok, sry if thats now to pendantic:
```
if (g_variant_check_format_string (data, "(suxxa(uus))", FALSE))
``` 
should be 
```
if (g_variant_check_format_string (data, "(suxxa(uus))", FALSE) && first s of variant == "GNOME")
```
since two portals could use the same variant format. And restoring someone else's session would most likely be just not restoring anything and querying the user again or adding support to parse another portal implementations?

So my two concerns are:
* Collisions in the format_variant (Can be mitigated by checking the format and every portal impl individually placing a magic token somewhere, therefore maybe standardizing on the position of this magic token on the first place)
* Loosing the state by switching DEs (Setting up a scene in obs in gnome, switching to kde, scene is started , kde can't parse the permissions and drops them, user sets up new permissions, switches back to gnome, all permissions get dropped ...) 

Since the application has to store the token and doesn't know about the used portal implementation the second point seems not solvable, except an application might check sth. like `XDG_CURRENT_DESKTOP` and store multiple tokens for each DE that way.

Please correct me if I got sth. wrong or am missunderstanding sth.

Reply:
To be clear, adding an extra string is necessarily a bad idea. The problem is where to draw the line. Is a string alone sufficient? Should it also be versioned? Who should set these strings? Are portal implementations allowed to "lie" about who their vendor strings? Should xdg-desktop-portal itself derive this vendor string? If so, where from?

These are all API questions that only exist if we try to add more data to this, and I'm reasonably sure that there is no decision that would make everyone happy. Ultimately, that's why I opted for a plain `v` variant there - it's none of x-d-p's business, and backends can still try to restore the data they want, if so they wish.

Reply:
If you're changing your entire desktop environment, restoring streams is the smallest of your problems. If you're on KDE and switch to GNOME, or to Sway, or to any other environment, there is very little to be carried over. This API is made for restoring streams, optimizing this API for this case is going to overcomplicate things with essentially no gains.

Reply:
> To be clear, adding an extra string is necessarily a bad idea. The problem is where to draw the line. Is a string alone sufficient? Should it also be versioned? Who should set these strings? Are portal implementations allowed to "lie" about their vendor strings? Should xdg-desktop-portal itself derive this vendor string? If so, where from?

Sry now I get the point of your missunderstanding. Against xdg-desktop-portal an arbitrary variant is really fine! What I meant was more an agreement between portal implementations to start the content of their variant with a "vendor name".

Reply:
> This API is made for restoring streams, optimizing this API for this case is going to overcomplicate things with essentially no gains.

And sry again for being out of the scope of this API discussion: We can keep the API simple as is and if applications want/have a need to work on mutliple DE's they won't be able to carry over their settings (I didn't want to imply that, since the DE's might be too different for that), but they can detect the running DE and only use tokens obtained while running with this DE and as such keep all permissions wrt. the running DE.

Reply:
>  What I meant was more an agreement between portal implementations to start the content of their variant with a "vendor name".

Oh I see. We can absolutely agree on a format with a at least few common fields that can be interpreted by every implementation. What I was trying to say is that I don't think this should be enforced in *here* - this API provides the mechanism, not the policy, to restore data.

Reply:
> We can absolutely agree on a format with a at least few common fields that can be interpreted by every implementation. What I was trying to say is that I don't think this should be enforced in here - this API provides the mechanism, not the policy, to restore data.

I think it's OK for the implementation to not try to validate (motto: "mechanism, not policy"), but it would still be useful to mention the convention in the specification.

It might be best to establish a convention that the `restore_data` is a struct starting with something like `GNOME-0` so that if a future `x-d-p-gnome` changes the information that it needs to store, it can use `GNOME-1`, ... for incompatible struct payloads? Or were you intending to use an append-only struct similar to the convention used in Linux syscalls?